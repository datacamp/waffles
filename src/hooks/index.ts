export { default as useId } from './use-id';
export { default as useMediaQuery } from './use-media-query';
