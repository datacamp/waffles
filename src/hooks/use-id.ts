import { useState, useEffect } from 'react';
import { nanoid } from 'nanoid';

/**
 * Autogenerate random ID to facilitate accessibility features.
 *
 * For a components which are part of server-side initial render it may be required to provide IDs manually to avoid client-server values mismatch.
 *
 * @param prefix Custom prefix for an ID
 * @returns Random ID with a prefix
 */
function useId(prefix: string) {
  const [id, setId] = useState(`${prefix}-0`);

  useEffect(() => {
    setId(nanoid(6));
  }, []);

  return `${prefix}-${id}`;
}

export default useId;
