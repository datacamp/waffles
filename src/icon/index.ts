export { default as AddCircleInverted } from './output/add-circle-inverted';
export { default as AddCircle } from './output/add-circle';
export { default as AddUser } from './output/add-user';
export { default as AppleBrand } from './output/apple-brand';
export { default as Apply } from './output/apply';
export { default as Archive } from './output/archive';
export { default as ArrowDown } from './output/arrow-down';
export { default as ArrowLeft } from './output/arrow-left';
export { default as ArrowRight } from './output/arrow-right';
export { default as ArrowUp } from './output/arrow-up';
export { default as Assess } from './output/assess';
export { default as AssignmentInverted } from './output/assignment-inverted';
export { default as Assignment } from './output/assignment';
export { default as Attachment } from './output/attachment';
export { default as AttentionCircleInverted } from './output/attention-circle-inverted';
export { default as AttentionCircle } from './output/attention-circle';
export { default as Back } from './output/back';
export { default as BarChart } from './output/bar-chart';
export { default as Bell } from './output/bell';
export { default as Book } from './output/book';
export { default as BookmarkInverted } from './output/bookmark-inverted';
export { default as Bookmark } from './output/bookmark';
export { default as Branch } from './output/branch';
export { default as Briefcase } from './output/briefcase';
export { default as Building } from './output/building';
export { default as Bulb } from './output/bulb';
export { default as Bullet } from './output/bullet';
export { default as Calendar } from './output/calendar';
export { default as Camera } from './output/camera';
export { default as ChatInverted } from './output/chat-inverted';
export { default as Chat } from './output/chat';
export { default as CheckmarkCircleInverted } from './output/checkmark-circle-inverted';
export { default as CheckmarkCircle } from './output/checkmark-circle';
export { default as Checkmark } from './output/checkmark';
export { default as ChevronDown } from './output/chevron-down';
export { default as ChevronLeft } from './output/chevron-left';
export { default as ChevronRight } from './output/chevron-right';
export { default as ChevronUp } from './output/chevron-up';
export { default as CircleInverted } from './output/circle-inverted';
export { default as Circle } from './output/circle';
export { default as ClockInverted } from './output/clock-inverted';
export { default as Clock } from './output/clock';
export { default as Code } from './output/code';
export { default as CogInverted } from './output/cog-inverted';
export { default as Cog } from './output/cog';
export { default as Columns } from './output/columns';
export { default as Command } from './output/command';
export { default as Compass } from './output/compass';
export { default as Copy } from './output/copy';
export { default as CreditCard } from './output/credit-card';
export { default as CrossCircleInverted } from './output/cross-circle-inverted';
export { default as CrossCircle } from './output/cross-circle';
export { default as Cross } from './output/cross';
export { default as Crosshair } from './output/crosshair';
export { default as DataCampBrand } from './output/data-camp-brand';
export { default as Database } from './output/database';
export { default as Desktop } from './output/desktop';
export { default as Diamond } from './output/diamond';
export { default as Document } from './output/document';
export { default as Dollar } from './output/dollar';
export { default as Download } from './output/download';
export { default as Edit } from './output/edit';
export { default as Education } from './output/education';
export { default as ExerciseInverted } from './output/exercise-inverted';
export { default as Exercise } from './output/exercise';
export { default as Exit } from './output/exit';
export { default as Expand } from './output/expand';
export { default as ExternalLink } from './output/external-link';
export { default as FacebookBrand } from './output/facebook-brand';
export { default as Feed } from './output/feed';
export { default as Filter } from './output/filter';
export { default as Flag } from './output/flag';
export { default as FlashInverted } from './output/flash-inverted';
export { default as Flash } from './output/flash';
export { default as Flashcard } from './output/flashcard';
export { default as Float } from './output/float';
export { default as Folder } from './output/folder';
export { default as Footer } from './output/footer';
export { default as Forbidden } from './output/forbidden';
export { default as Forward } from './output/forward';
export { default as FullScreen } from './output/full-screen';
export { default as GithubBrand } from './output/github-brand';
export { default as Globe } from './output/globe';
export { default as GoogleBrand } from './output/google-brand';
export { default as GoogleColoredBrand } from './output/google-colored-brand';
export { default as GooglePlayBrand } from './output/google-play-brand';
export { default as GroupInverted } from './output/group-inverted';
export { default as Group } from './output/group';
export { default as Header } from './output/header';
export { default as HealthInverted } from './output/health-inverted';
export { default as Health } from './output/health';
export { default as HeartInverted } from './output/heart-inverted';
export { default as Heart } from './output/heart';
export { default as HelpCircleInverted } from './output/help-circle-inverted';
export { default as HelpCircle } from './output/help-circle';
export { default as Help } from './output/help';
export { default as HiddenInverted } from './output/hidden-inverted';
export { default as Hidden } from './output/hidden';
export { default as Home } from './output/home';
export { default as Image } from './output/image';
export { default as InfoCircleInverted } from './output/info-circle-inverted';
export { default as InfoCircle } from './output/info-circle';
export { default as InstagramBrand } from './output/instagram-brand';
export { default as Integration } from './output/integration';
export { default as Interface } from './output/interface';
export { default as Introduction } from './output/introduction';
export { default as KaggleBrand } from './output/kaggle-brand';
export { default as Keyboard } from './output/keyboard';
export { default as Lab } from './output/lab';
export { default as LeanpubBrand } from './output/leanpub-brand';
export { default as Learn } from './output/learn';
export { default as LineChart } from './output/line-chart';
export { default as Link } from './output/link';
export { default as LinkedInBrand } from './output/linked-in-brand';
export { default as List } from './output/list';
export { default as LockedInverted } from './output/locked-inverted';
export { default as Locked } from './output/locked';
export { default as Mail } from './output/mail';
export { default as Markdown } from './output/markdown';
export { default as Matrix } from './output/matrix';
export { default as Medal } from './output/medal';
export { default as Menu } from './output/menu';
export { default as MicrophoneInverted } from './output/microphone-inverted';
export { default as Microphone } from './output/microphone';
export { default as Mobile } from './output/mobile';
export { default as MoonInverted } from './output/moon-inverted';
export { default as Moon } from './output/moon';
export { default as More } from './output/more';
export { default as MuteInverted } from './output/mute-inverted';
export { default as Mute } from './output/mute';
export { default as NineDots } from './output/nine-dots';
export { default as Pause } from './output/pause';
export { default as PaypalBrand } from './output/paypal-brand';
export { default as PaypalColoredBrand } from './output/paypal-colored-brand';
export { default as Phone } from './output/phone';
export { default as PieChart } from './output/pie-chart';
export { default as PinInverted } from './output/pin-inverted';
export { default as Pin } from './output/pin';
export { default as PlatformSwitch } from './output/platform-switch';
export { default as PlayInverted } from './output/play-inverted';
export { default as Play } from './output/play';
export { default as Plus } from './output/plus';
export { default as Practice } from './output/practice';
export { default as Publish } from './output/publish';
export { default as PythonBrand } from './output/python-brand';
export { default as RBrand } from './output/r-brand';
export { default as Random } from './output/random';
export { default as Redo } from './output/redo';
export { default as Refresh } from './output/refresh';
export { default as RemoveInverted } from './output/remove-inverted';
export { default as Remove } from './output/remove';
export { default as Reply } from './output/reply';
export { default as RewindInverted } from './output/rewind-inverted';
export { default as Rewind } from './output/rewind';
export { default as RocketInverted } from './output/rocket-inverted';
export { default as Rocket } from './output/rocket';
export { default as Rows } from './output/rows';
export { default as Save } from './output/save';
export { default as ScalaBrand } from './output/scala-brand';
export { default as Search } from './output/search';
export { default as Share } from './output/share';
export { default as SkipInverted } from './output/skip-inverted';
export { default as Skip } from './output/skip';
export { default as SlackBrand } from './output/slack-brand';
export { default as SlackColoredBrand } from './output/slack-colored-brand';
export { default as SoundInverted } from './output/sound-inverted';
export { default as Sound } from './output/sound';
export { default as SparklesInverted } from './output/sparkles-inverted';
export { default as Sparkles } from './output/sparkles';
export { default as StarInverted } from './output/star-inverted';
export { default as Star } from './output/star';
export { default as StopInverted } from './output/stop-inverted';
export { default as Stop } from './output/stop';
export { default as SunInverted } from './output/sun-inverted';
export { default as Sun } from './output/sun';
export { default as Tab } from './output/tab';
export { default as Tag } from './output/tag';
export { default as Team } from './output/team';
export { default as TerminalInverted } from './output/terminal-inverted';
export { default as Terminal } from './output/terminal';
export { default as Text } from './output/text';
export { default as TrashInverted } from './output/trash-inverted';
export { default as Trash } from './output/trash';
export { default as TrimBegin } from './output/trim-begin';
export { default as TrimEnd } from './output/trim-end';
export { default as Trophy } from './output/trophy';
export { default as TwitterBrand } from './output/twitter-brand';
export { default as Undo } from './output/undo';
export { default as Unlink } from './output/unlink';
export { default as UnlockedInverted } from './output/unlocked-inverted';
export { default as Unlocked } from './output/unlocked';
export { default as Upload } from './output/upload';
export { default as UpvoteInverted } from './output/upvote-inverted';
export { default as Upvote } from './output/upvote';
export { default as UserInverted } from './output/user-inverted';
export { default as User } from './output/user';
export { default as VerifiedInverted } from './output/verified-inverted';
export { default as Verified } from './output/verified';
export { default as VideoCameraInverted } from './output/video-camera-inverted';
export { default as VideoCamera } from './output/video-camera';
export { default as VideoInverted } from './output/video-inverted';
export { default as Video } from './output/video';
export { default as VisibleInverted } from './output/visible-inverted';
export { default as Visible } from './output/visible';
export { default as Waypoint } from './output/waypoint';
export { default as YoutubeBrand } from './output/youtube-brand';
