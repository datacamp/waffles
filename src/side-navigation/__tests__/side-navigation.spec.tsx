/* eslint-disable @typescript-eslint/no-empty-function */
import { useRef, useEffect } from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react';

import { Star, ChevronRight } from '../../icon';
import { useMediaQuery } from '../../hooks';
import { SideNavigation } from '../index';

const MOCKED_ID = '123abC';

jest.mock('../../hooks', () => {
  return {
    ...jest.requireActual('../../hooks'),
    useMediaQuery: jest.fn(),
    useId: (prefix: string) => {
      return `${prefix}-${MOCKED_ID}`;
    },
  };
});

const useMediaQueryMock = useMediaQuery as jest.Mock;

type TestSideNavProps = {
  isOpen: boolean;
  onClose: () => void;
};

function TestSideNav({ isOpen, onClose }: TestSideNavProps) {
  return (
    <SideNavigation isOpen={isOpen} onClose={onClose}>
      <SideNavigation.Nav>
        <SideNavigation.Category label="Trending" data-testid="main-category">
          <SideNavigation.Item href="#taylor" isActive>
            Taylor Swift
          </SideNavigation.Item>
          <SideNavigation.Item href="#ariana" isNew>
            Ariana Grande
          </SideNavigation.Item>
          <SideNavigation.Item
            href="#justin"
            iconLeft={<Star data-testid="left-icon" />}
            iconRight={<ChevronRight data-testid="right-icon" />}
          >
            Justin Bieber
          </SideNavigation.Item>
          <SideNavigation.Subcategory label="Classics">
            <SideNavigation.Item href="#led-zeppelin">
              Led Zeppelin
            </SideNavigation.Item>
            <SideNavigation.Item
              href="#pink-floyd"
              iconLeft={<Star data-testid="sub-item-left-icon" />}
              iconRight={<ChevronRight data-testid="sub-item-right-icon" />}
            >
              Pink Floyd
            </SideNavigation.Item>
          </SideNavigation.Subcategory>
          <SideNavigation.Subcategory
            label="New Age"
            iconLeft={<Star data-testid="sub-left-icon" />}
            iconRight={<ChevronRight data-testid="sub-right-icon" />}
          >
            <SideNavigation.Item href="#greta-van-fleet">
              Greta Van Fleet
            </SideNavigation.Item>
          </SideNavigation.Subcategory>
        </SideNavigation.Category>
      </SideNavigation.Nav>
    </SideNavigation>
  );
}

type TestComponentProps = {
  to: string;
} & React.ComponentPropsWithoutRef<'a'>;

function TestComponent({ children, to, ...restProps }: TestComponentProps) {
  return (
    <a href={to} {...restProps}>
      {children}
    </a>
  );
}

function PolymorphicTestSideNav({ isOpen, onClose }: TestSideNavProps) {
  return (
    <>
      <SideNavigation isOpen={isOpen} onClose={onClose}>
        <SideNavigation.Nav>
          <SideNavigation.Category label="Trending">
            <SideNavigation.Item as={TestComponent} to="#taylor" isActive>
              Taylor Swift
            </SideNavigation.Item>
            <SideNavigation.Subcategory
              label="Classics"
              as={TestComponent}
              to="#classics"
            >
              <SideNavigation.Item as="button">
                Led Zeppelin
              </SideNavigation.Item>
            </SideNavigation.Subcategory>
          </SideNavigation.Category>
        </SideNavigation.Nav>
      </SideNavigation>
    </>
  );
}

function TestRefItem() {
  const itemRef = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    itemRef.current?.focus();
  }, []);

  return (
    <SideNavigation.Item ref={itemRef} href="/test">
      Focused With Ref
    </SideNavigation.Item>
  );
}

describe('SideNavigation', () => {
  beforeEach(() => {
    jest.resetAllMocks();

    useMediaQueryMock.mockReturnValue({
      isAboveSmall: true,
      isAboveMedium: false,
      isAboveLarge: false,
    });
  });

  it('category and its items are associacted by same ID', () => {
    const { getByTestId } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const category = getByTestId('main-category');
    const categoryLabel = category.querySelector('h2');
    const list = category.querySelector('ul');
    const categoryId = `menu-category-${MOCKED_ID}`;

    expect(categoryLabel).toHaveAttribute('id', categoryId);
    expect(list).toHaveAttribute('aria-labelledby', categoryId);
  });

  it('subcategory and its items are associacted by same ID', () => {
    const { getByText } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const subcategoryLabel = getByText('Classics').closest('h2');
    const list = subcategoryLabel?.closest('li')?.querySelector('ul');
    const categoryId = `menu-subcategory-${MOCKED_ID}`;

    expect(subcategoryLabel).toHaveAttribute('id', categoryId);
    expect(list).toHaveAttribute('aria-labelledby', categoryId);
  });

  it('active link has correct aria attribute', () => {
    const { getByText } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const activeLink = getByText('Taylor Swift').closest('a');

    expect(activeLink).toHaveAttribute('href', '#taylor');
    expect(activeLink).toHaveAttribute('aria-current', 'page');
  });

  it('nav item could be rendered as a button', () => {
    const { container } = render(
      <PolymorphicTestSideNav isOpen={true} onClose={() => {}} />,
    );

    const item = container.querySelector('button');
    expect(item).toHaveTextContent('Led Zeppelin');
  });

  it('nav item requires props owned by custom component passed into "as" prop', () => {
    const { getByText } = render(
      <PolymorphicTestSideNav isOpen={true} onClose={() => {}} />,
    );

    const item = getByText('Taylor Swift').closest('a');
    expect(item).toHaveAttribute('href', '#taylor');
  });

  it('nav subcategory requires props owned by custom component passed into "as" prop', () => {
    const { getByText } = render(
      <PolymorphicTestSideNav isOpen={true} onClose={() => {}} />,
    );

    const subcategoryAsLink = getByText('Classics').closest('a');
    expect(subcategoryAsLink).toHaveAttribute('href', '#classics');
  });

  it('nav subcategory could be decorated with left and right icon', () => {
    const { getByText, getByTestId } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const subcategoryWithIcons = getByText('New Age').closest('h2');
    const leftIcon = getByTestId('sub-left-icon');
    const rightIcon = getByTestId('sub-right-icon');

    expect(subcategoryWithIcons).toContainElement(leftIcon);
    expect(subcategoryWithIcons).toContainElement(rightIcon);
  });

  it('nav item could be decorated with left and right icon', () => {
    const { getByText, getByTestId } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const itemWithIcons = getByText('Justin Bieber').closest('a');
    const leftIcon = getByTestId('left-icon');
    const rightIcon = getByTestId('right-icon');

    expect(itemWithIcons).toContainElement(leftIcon);
    expect(itemWithIcons).toContainElement(rightIcon);
  });

  it('for subcategory nav item only right icon is visible', () => {
    const { getByText, getByTestId, queryByTestId } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const itemWithIcons = getByText('Pink Floyd').closest('a');
    const leftIcon = queryByTestId('sub-item-left-icon');
    const rightIcon = getByTestId('sub-item-right-icon');

    expect(itemWithIcons).toContainElement(rightIcon);
    expect(leftIcon).not.toBeInTheDocument();
  });

  it('when item is new, show appropriate indicator', () => {
    const { getByText, getByTestId } = render(
      <TestSideNav isOpen={true} onClose={() => {}} />,
    );

    const newItem = getByText('Ariana Grande').closest('a');
    const indicator = getByTestId('new-item-indicator');

    expect(newItem).toContainElement(indicator);
  });

  // Mobile
  describe('below small breakpoint', () => {
    beforeEach(() => {
      jest.useFakeTimers();
      jest.resetAllMocks();

      useMediaQueryMock.mockReturnValue({
        isAboveSmall: false,
        isAboveMedium: false,
        isAboveLarge: false,
      });
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('is not visible when isOpen is set to false', () => {
      const { container, queryByTestId } = render(
        <TestSideNav isOpen={false} onClose={() => {}} />,
      );

      const nav = container.querySelector('nav');
      const overlay = queryByTestId('side-navigation-menu-overlay');

      expect(nav).not.toBeInTheDocument();
      expect(overlay).not.toBeInTheDocument();
    });

    it('is visible when isOpen is set to true', async () => {
      const { getByTestId, getByLabelText } = render(
        <TestSideNav isOpen={true} onClose={() => {}} />,
      );

      // Let all animations finish
      act(() => {
        jest.advanceTimersByTime(500);
      });

      let overlay;
      await waitFor(() => {
        overlay = getByTestId('side-navigation-menu-overlay');
      });
      const closeButton = getByLabelText('Close');

      expect(overlay).toBeInTheDocument();
      expect(closeButton).toBeInTheDocument();
    });

    it('when close button is clicked, onClose handler is called', async () => {
      const handleOnClose = jest.fn();
      const { getByLabelText } = render(
        <TestSideNav isOpen={true} onClose={handleOnClose} />,
      );

      act(() => {
        jest.advanceTimersByTime(500);
      });

      let closeButton;
      await waitFor(() => {
        closeButton = getByLabelText('Close');
      });
      closeButton && fireEvent.click(closeButton);

      expect(handleOnClose).toHaveBeenCalledTimes(1);
    });

    it('when nav item is clicked, onClose handler is called', async () => {
      const handleOnClose = jest.fn();
      const { getByText } = render(
        <TestSideNav isOpen={true} onClose={handleOnClose} />,
      );

      act(() => {
        jest.advanceTimersByTime(500);
      });

      let navItem;
      await waitFor(() => {
        navItem = getByText('Ariana Grande').closest('a');
      });
      navItem && fireEvent.click(navItem);

      expect(handleOnClose).toHaveBeenCalledTimes(1);
    });

    it('when overlay is clicked, onClose handler is called', async () => {
      const handleOnClose = jest.fn();
      const { getByTestId } = render(
        <TestSideNav isOpen={true} onClose={handleOnClose} />,
      );

      act(() => {
        jest.advanceTimersByTime(500);
      });

      let overlay;
      await waitFor(() => {
        overlay = getByTestId('side-navigation-menu-overlay');
      });
      overlay && fireEvent.click(overlay);

      expect(handleOnClose).toHaveBeenCalledTimes(1);
    });
  });

  // Tablet & Desktop
  describe('above small breakpoint', () => {
    beforeEach(() => {
      jest.resetAllMocks();

      useMediaQueryMock.mockReturnValue({
        isAboveSmall: true,
        isAboveMedium: true,
        isAboveLarge: true,
      });
    });

    it('is always visible, even when isOpen is set to false', () => {
      const { container, queryByTestId, queryByLabelText } = render(
        <TestSideNav isOpen={false} onClose={() => {}} />,
      );

      const nav = container.querySelector('nav');
      const categories = container.querySelectorAll('h2');
      const links = container.querySelectorAll('a');
      const overlay = queryByTestId('side-navigation-menu-overlay');
      const closeButton = queryByLabelText('Close');

      expect(nav).toBeInTheDocument();
      expect(categories).toHaveLength(3);
      expect(links).toHaveLength(6);
      expect(overlay).not.toBeInTheDocument();
      expect(closeButton).not.toBeInTheDocument();
    });

    it('when nav item is clicked, onClose handler is NOT called', () => {
      const handleOnClose = jest.fn();
      const { getByText } = render(
        <TestSideNav isOpen={true} onClose={handleOnClose} />,
      );

      const navItem = getByText('Ariana Grande').closest('a');
      navItem && fireEvent.click(navItem);

      expect(handleOnClose).toHaveBeenCalledTimes(0);
    });
  });

  it('for nav item click is handled correctly', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <SideNavigation.Item href="#taylor" onClick={handleClick}>
        Taylor Swift
      </SideNavigation.Item>,
    );

    const item = getByText('Taylor Swift').closest('a');
    item && fireEvent.click(item);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('for nav subcategory click is handled correctly', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <SideNavigation.Subcategory
        label="Trending"
        as="a"
        href="#taylor"
        onClick={handleClick}
      >
        Test
      </SideNavigation.Subcategory>,
    );

    const item = getByText('Trending').closest('a');
    item && fireEvent.click(item);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('for nav item focus is handled correctly', () => {
    const handleFocus = jest.fn();
    const { getByText } = render(
      <SideNavigation.Item href="#taylor" onFocus={handleFocus}>
        Taylor Swift
      </SideNavigation.Item>,
    );

    const item = getByText('Taylor Swift').closest('a');
    item && fireEvent.focus(item);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('for nav subcategory focus is handled correctly', () => {
    const handleFocus = jest.fn();
    const { getByText } = render(
      <SideNavigation.Subcategory
        label="Trending"
        as="a"
        href="#taylor"
        onFocus={handleFocus}
      >
        Test
      </SideNavigation.Subcategory>,
    );

    const item = getByText('Trending').closest('a');
    item && fireEvent.focus(item);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('nav item accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefItem />);

    const item = container.querySelector('a');

    expect(item).toHaveFocus();
  });

  // Snapshots

  it('renders snapshot of category with label and divider', () => {
    const { container } = render(
      <SideNavigation.Category label="Category">Test</SideNavigation.Category>,
    );

    const category = container.firstChild;

    expect(category).toMatchSnapshot();
  });

  it('renders snapshot of category without label nor divider', () => {
    const { container } = render(
      <SideNavigation.Category>Test</SideNavigation.Category>,
    );

    const category = container.firstChild;

    expect(category).toMatchSnapshot();
  });

  it('renders snapshot of the small item', () => {
    const { container } = render(
      <SideNavigation.Item size="small" href="#test">
        Small
      </SideNavigation.Item>,
    );

    const category = container.querySelector('li');

    expect(category).toMatchSnapshot();
  });

  it('renders snapshot of the medium item', () => {
    const { container } = render(
      <SideNavigation.Item size="medium" href="#test">
        Medium
      </SideNavigation.Item>,
    );

    const category = container.querySelector('li');

    expect(category).toMatchSnapshot();
  });

  it('renders snapshot of item focused state', () => {
    const { getByText } = render(
      <SideNavigation.Item href="#test">Focused</SideNavigation.Item>,
    );

    const item = getByText('Focused').closest('a') as HTMLAnchorElement;
    fireEvent.focus(item);

    expect(item).toMatchSnapshot();
  });
});
