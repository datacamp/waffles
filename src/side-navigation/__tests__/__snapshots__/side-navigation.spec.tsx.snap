// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigation renders snapshot of category with label and divider 1`] = `
.emotion-0 {
  height: 1px;
  background-color: rgba(255, 255, 255, 0.2);
  margin-top: 8px;
  margin-bottom: 8px;
  margin-left: 16px;
  margin-right: 16px;
}

.emotion-1 {
  color: #FFFFFF;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 12px;
  font-weight: 800;
  line-height: 1.25;
  letter-spacing: 1.5px;
  text-transform: uppercase;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  margin: 0;
}

.emotion-2 {
  margin: 0;
  padding: 0;
  list-style: none;
}

<div>
  <div
    class="emotion-0"
  />
  <h2
    class="emotion-1"
    id="menu-category-123abC"
  >
    Category
  </h2>
  <ul
    aria-labelledby="menu-category-123abC"
    class="emotion-2"
  >
    Test
  </ul>
</div>
`;

exports[`SideNavigation renders snapshot of category without label nor divider 1`] = `
.emotion-0 {
  height: 1px;
  background-color: rgba(255, 255, 255, 0.2);
  margin-top: 8px;
  margin-bottom: 8px;
  margin-left: 16px;
  margin-right: 16px;
}

.emotion-1 {
  margin: 0;
  padding: 0;
  list-style: none;
}

<div>
  <div
    class="emotion-0"
  />
  <ul
    class="emotion-1"
  >
    Test
  </ul>
</div>
`;

exports[`SideNavigation renders snapshot of item focused state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #FFFFFF;
  background-color: transparent;
  border: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 16px;
  padding-right: 16px;
  margin: 0;
  opacity: 0.8;
  outline: 0;
  width: 100%;
  -webkit-transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  box-shadow: inset 0 0 0 2px #009BD8;
}

.emotion-0 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:where(a, button) {
  cursor: pointer;
}

@media screen and (min-width: 480px) {
  .emotion-0:where(a, button):hover {
    opacity: 1;
    background-color: rgba(255, 255, 255, 0.1);
  }
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
  color: #FFFFFF;
  line-height: 1.25;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-size: 16px;
  min-height: 42px;
  padding-top: 8px;
  padding-bottom: 8px;
}

@media screen and (min-width: 480px) {
  .emotion-1 {
    font-size: 14px;
    min-height: 36px;
    padding-top: 8px;
    padding-bottom: 8px;
  }
}

<a
  class="emotion-0"
  href="#test"
>
  <span
    class="emotion-1"
  >
    Focused
  </span>
</a>
`;

exports[`SideNavigation renders snapshot of the medium item 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #FFFFFF;
  background-color: transparent;
  border: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 16px;
  padding-right: 16px;
  margin: 0;
  opacity: 0.8;
  outline: 0;
  width: 100%;
  -webkit-transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:where(a, button) {
  cursor: pointer;
}

@media screen and (min-width: 480px) {
  .emotion-0:where(a, button):hover {
    opacity: 1;
    background-color: rgba(255, 255, 255, 0.1);
  }
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
  color: #FFFFFF;
  line-height: 1.25;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-size: 16px;
  min-height: 42px;
  padding-top: 8px;
  padding-bottom: 8px;
}

@media screen and (min-width: 480px) {
  .emotion-1 {
    font-size: 14px;
    min-height: 36px;
    padding-top: 8px;
    padding-bottom: 8px;
  }
}

<li>
  <a
    class="emotion-0"
    href="#test"
  >
    <span
      class="emotion-1"
    >
      Medium
    </span>
  </a>
</li>
`;

exports[`SideNavigation renders snapshot of the small item 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #FFFFFF;
  background-color: transparent;
  border: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 16px;
  padding-right: 16px;
  margin: 0;
  opacity: 0.8;
  outline: 0;
  width: 100%;
  -webkit-transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  transition: background-color 75ms ease-out,color 75ms ease-out,box-shadow 125ms ease-out;
  cursor: default;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-0:where(a, button) {
  cursor: pointer;
}

@media screen and (min-width: 480px) {
  .emotion-0:where(a, button):hover {
    opacity: 1;
    background-color: rgba(255, 255, 255, 0.1);
  }
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: left;
  color: #FFFFFF;
  line-height: 1.25;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-size: 14px;
  min-height: 36px;
  padding-top: 4px;
  padding-bottom: 4px;
}

@media screen and (min-width: 480px) {
  .emotion-1 {
    font-size: 12px;
    min-height: 28px;
    padding-top: 4px;
    padding-bottom: 4px;
  }
}

<li>
  <a
    class="emotion-0"
    href="#test"
  >
    <span
      class="emotion-1"
    >
      Small
    </span>
  </a>
</li>
`;
